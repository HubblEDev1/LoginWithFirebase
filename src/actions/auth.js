import { createUserWithEmailAndPassword, getAuth, signInWithEmailAndPassword, signInWithPopup, signOut, updateProfile } from "@firebase/auth";
import Swal from "sweetalert2";
import { googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types"
import { finishLoading, startLoading } from "./ui";

/*export const login = (uid, displayName) => {
    return {
        type: types.login
    }
}*/
export const startLoginEmailPasswords = (email, password) => {
    return (dispatch) => {
        /*setTimeout(() => {
            dispatch(login(123, 'edwin'))
        }, 3500);*/
        dispatch(startLoading());
        const auth = getAuth();
        signInWithEmailAndPassword(auth, email, password).then( ({user}) => {
            console.log(user);
            dispatch(finishLoading());
            dispatch(login(user.uid, user.displayName));
        }).catch((e) => {
            console.log(e);
            dispatch(finishLoading());
            Swal.fire('Error', e.message, 'error');
        })
    }
}

export const startGoogleLogin = () => {
    return ( dispatch ) => {
        const auth = getAuth();
        signInWithPopup(auth, googleAuthProvider).then(({user}) => {
            dispatch(login(user.uid, user.displayName));
        });
    }
}

export const startRegisterWithEmailPasswordName = (email, pass, name) => {
    return (dispatch) => {
        const auth = getAuth();
        createUserWithEmailAndPassword(auth, email, pass).then( async ({user}) => {
            await updateProfile(auth.currentUser, {displayName: name});
            dispatch(
                login(user.uid, user.displayName)
            );
        }).catch( e => {
            console.log(e);
            Swal.fire('Error', e.message, 'error');
        })
    }
}


export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid, 
        displayName
    }
}); 

export const startLogout = () => {
    return async ( dispatch ) => {
        const auth = getAuth();
        signOut(auth);
        dispatch( logout() );
    }
}

export const logout = () => ({
    type: types.logout,
})