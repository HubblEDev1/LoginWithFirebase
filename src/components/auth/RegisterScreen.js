import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm';

import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen = () => {
    const dispatch = useDispatch();
    //const state = useSelector( (state) => state);
    const { msgError } = useSelector( state => state.ui );

    const [formValues, handleInputChange, reset] = useForm({
        name: '',
        email: '',
        pass: '',
        pass2: ''
    });

    const {name, email, pass, pass2} = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        if( isFormValid() ){
            dispatch(startRegisterWithEmailPasswordName(email, pass, name));
        }
    }

    const isFormValid = () => {

        if(name.trim().length === 0){
            dispatch(setError('Name is required'));
            return false;
        }else if(!validator.isEmail(email)){
            dispatch(setError('Email is not valid'))
            return false;
        }else if( pass !== pass2 || pass < 5 ){
            dispatch(setError('Incorrect password'));
            return false;
        }

        dispatch(removeError());
        
        return true;
    }

    return (
        <>
            <h3 className="auth__title">Register</h3>
            <form onSubmit={handleRegister}>
                {
                    msgError && (
                        <div className="auth__alert-error">
                            {msgError}
                        </div>
                    )
                }
                <input 
                    type="text"
                    placeholder="Name" 
                    name="name"
                    className="auth__input" 
                    value={name}    
                    onChange={handleInputChange}           
                />
                <input 
                    type="text"
                    placeholder="Email" 
                    name="email"
                    className="auth__input"
                    value={email} 
                    onChange={handleInputChange}                 
                />
                <input 
                    type="password"
                    placeholder="Password" 
                    name="pass"
                    className="auth__input"
                    value={pass} 
                    onChange={handleInputChange} 
                />
                <input 
                    type="password"
                    placeholder="Confirm password" 
                    name="pass2"
                    className="auth__input"
                    value={pass2} 
                    onChange={handleInputChange} 
                />
                <button
                    type="submit"
                    className="btn btn-primary btn-block mb-5" 
                >
                    Register
                </button>


                <Link
                    to="/auth/login"
                    className="link"
                >
                    Already Registered?
                </Link>
            </form>
        </>
    )
}
