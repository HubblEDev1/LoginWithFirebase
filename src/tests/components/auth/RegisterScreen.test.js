import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';
import '../../../setupTests';

import { mount } from "enzyme";
import { Provider } from "react-redux";
import { LoginScreen } from "../../../components/auth/LoginScreen"
import { MemoryRouter } from 'react-router';
import { startGoogleLogin, startLoginEmailPasswords } from '../../../actions/auth';
import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';

jest.mock('../../../actions/auth', () => ({
    startGoogleLogin: jest.fn(),
    startLoginEmailPasswords: jest.fn(),
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
};

let store = mockStore(initState);
//store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store = { store }>
        <MemoryRouter>
            <RegisterScreen/>
        </MemoryRouter>
    </Provider>
);

describe('Testing RegisterScreen/>', () => {
    /*
    beforeEach( () => {
        store = mockStore(initState);
        jest.clearAllMocks();
    } );*/

    test('should snapshot ', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should dispatch of the action', () => {
        const emailField = wrapper.find('input[name="email"]');
        //console.log(emailField.exists());
        emailField.simulate('change', {
            target: {
                value: '',
                name: 'email'
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        const actions = store.getActions();
        console.log(actions);
        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'Name is required'
        })
    });

    test('should show message', () => {
        const initState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'Email is not correct'
            }
        };
        
        const store = mockStore(initState);
        //store.dispatch = jest.fn();
        const wrapper = mount(
            <Provider store = { store }>
                <MemoryRouter>
                    <RegisterScreen/>
                </MemoryRouter>
            </Provider>
        );

        expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe( initState.ui.msgError );
        
    })
    


    
    /*
    test('should dispatch googleLogin ', () => {
        wrapper.find('.google-btn').prop('onClick')();

        expect(startGoogleLogin).toHaveBeenCalled();
    });

    test('should dispatch startLoginEmailPassword', () => {
        wrapper.find('form').prop('onSubmit')({preventDefault(){}});
        expect(startLoginEmailPasswords).toHaveBeenCalledWith('edwin@example.com', '123456');
    })*/
});
