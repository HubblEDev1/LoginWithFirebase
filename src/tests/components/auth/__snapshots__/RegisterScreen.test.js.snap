// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen/> should snapshot  1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn"
    onSubmit={[Function]}
  >
    <input
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      className="auth__input"
      name="pass"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      className="auth__input"
      name="pass2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already Registered?
    </a>
  </form>,
]
`;

exports[`Testing RegisterScreen/> should snapshot  1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn"
    onSubmit={[Function]}
  >
    <input
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      className="auth__input"
      name="pass"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      className="auth__input"
      name="pass2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already Registered?
    </a>
  </form>,
]
`;
